pkg/providers/providers.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2024
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package providers
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	hmc "github.com/Mirantis/hmc/api/v1alpha1"
+	"github.com/Mirantis/hmc/pkg/credspropagation"
+	"github.com/Mirantis/hmc/pkg/providers/aws"
+)
+
+const (
+	InfraPrefix    = "infrastructure-"
+	ProviderPrefix = "cluster-api-provider-"
+)
+
+var (
+	mu sync.RWMutex
+
+	providers = []hmc.Provider{
+		{
+			Name: hmc.ProviderK0smotronName,
+		},
+		{
+			Name: hmc.ProviderSveltosName,
+		},
+	}
+
+	registry map[string]ProviderModule
+)
+
+type ProviderModule interface {
+	GetName() string
+	GetTitleName() string
+	GetClusterGVK() schema.GroupVersionKind
+	GetClusterIdentityKind() []string
+	CredentialPropagationFunc() func(
+		ctx context.Context,
+		propnCfg *credspropagation.PropagationCfg,
+		l logr.Logger,
+	) (enabled bool, err error)
+}
+
+func Register(p ProviderModule) {
+	mu.Lock()
+	defer mu.Unlock()
+
+	if registry == nil {
+		registry = make(map[string]ProviderModule)
+	}
+
+	shortName := p.GetName()
+
+	if _, exists := registry[shortName]; exists {
+		panic(fmt.Sprintf("provider %q already registered", shortName))
```

Comment:
I don't think so, error is basically about trying to register same provider twice, so this can be some conflicting name out-of-tree module, allowing provider to work will probably hide away expectations.

Also this function is called from `init()` of modules that we are not controlling (out-of-tree), so nowhere ti return error to.

Reply:
The suggestion was to convert the function from being noisy to be silent (quiet), which is probably against go-style patterns. But it is a matter of preference, I believe, so it is up to the implementer

Reply:
As an external user I would want to understand clearly why out-of-tree provider does not work, without reading code and placing debug prints.

External user can be someone that just uses some one else out-of-tree provider.